pkg/sdkclient/batchmapper/client.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package batchmapper
+
+import (
+	"context"
+	"errors"
+	"io"
+
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	batchmappb "github.com/numaproj/numaflow-go/pkg/apis/proto/batchmap/v1"
+	"github.com/numaproj/numaflow-go/pkg/info"
+
+	"github.com/numaproj/numaflow/pkg/sdkclient"
+	sdkerr "github.com/numaproj/numaflow/pkg/sdkclient/error"
+	grpcutil "github.com/numaproj/numaflow/pkg/sdkclient/grpc"
+)
+
+// client contains the grpc connection and the grpc client.
+type client struct {
+	conn    *grpc.ClientConn
+	grpcClt batchmappb.BatchMapClient
+}
+
+// New creates a new client object.
+func New(serverInfo *info.ServerInfo, inputOptions ...sdkclient.Option) (Client, error) {
+	var opts = sdkclient.DefaultOptions(sdkclient.BatchMapAddr)
+
+	for _, inputOption := range inputOptions {
+		inputOption(opts)
+	}
+
+	// Connect to the server
+	conn, err := grpcutil.ConnectToServer(opts.UdsSockAddr(), serverInfo, opts.MaxMessageSize())
+	if err != nil {
+		return nil, err
+	}
+
+	c := new(client)
+	c.conn = conn
+	c.grpcClt = batchmappb.NewBatchMapClient(conn)
+	return c, nil
+}
+
+func NewFromClient(c batchmappb.BatchMapClient) (Client, error) {
+	return &client{
+		grpcClt: c,
+	}, nil
+}
+
+// CloseConn closes the grpc client connection.
+func (c *client) CloseConn(ctx context.Context) error {
+	return c.conn.Close()
+}
+
+// IsReady returns true if the grpc connection is ready to use.
+func (c *client) IsReady(ctx context.Context, in *emptypb.Empty) (bool, error) {
+	resp, err := c.grpcClt.IsReady(ctx, in)
+	if err != nil {
+		return false, err
+	}
+	return resp.GetReady(), nil
+}
+
+// BatchMapFn is the handler for the gRPC client (Numa container)
+// It takes in a stream of input Requests, sends them to the gRPC server(UDF) and then streams the
+// responses received back on a channel asynchronously.
+// We spawn 2 goroutines here, one for sending the requests over the stream
+// and the other one for receiving the responses
+func (c *client) BatchMapFn(ctx context.Context, inputCh <-chan *batchmappb.BatchMapRequest) (<-chan *batchmappb.BatchMapResponse, <-chan error) {
+	// errCh is used to track and propagate any errors that might occur during the rpc lifecyle, these could include
+	// errors in sending, UDF errors etc
+	// These are propagated to the applier for further handling
+	errCh := make(chan error)
```

Comment:
who is closing this?

Reply:
For error channel we explicitly do not need to tell the reader that the channel has no more data. In that case this should be garbage collected when the routine ends right?

Or if its better to close I think we can add a close at
https://github.com/numaproj/numaflow/pull/1778/files/6b8271529878d5994cf1b20ffa15f09b9b5f4448#diff-bff711417b7f070d80a34939e929d12d12667a29719cc68e23671f3306a223daR117

Reply:
No, we can't close the error channel, because you are using select on the error channel and the response channel on the read side. So, if you close the error channel, it will get selected and there are chances that you never read from the response channel. So, in this case, we should set `errCh` to null. That way, the error channel will never get selected.